File: demos/hello_clearpath/hello_clearpath/hello_square.py
"""
Drive the robot in a small square using odometry feedback.

Example of a demo.
Python counterpart to the C++ ``hello_square_cpp`` executable. Use either to
verify your environment is set up correctly before trying more complex demos.
"""

import math

from geometry_msgs.msg import TwistStamped
from nav_msgs.msg import Odometry
import rclpy
from rclpy.node import Node

SIDE_LENGTH_M = 1.0
TURN_ANGLE_RAD = math.pi / 2
LINEAR_SPEED = 0.2  # m/s
ANGULAR_SPEED = 0.5  # rad/s
POSITION_TOLERANCE_M = 0.02
HEADING_TOLERANCE_RAD = math.radians(2.0)


def yaw_from_quaternion(q) -> float:
    """Extract yaw (rotation about Z) from a geometry_msgs Quaternion."""
    siny_cosp = 2.0 * (q.w * q.z + q.x * q.y)
    cosy_cosp = 1.0 - 2.0 * (q.y * q.y + q.z * q.z)
    return math.atan2(siny_cosp, cosy_cosp)


def angle_diff(a: float, b: float) -> float:
    """Return the smallest signed difference ``a - b`` wrapped to [-pi, pi]."""
    return math.atan2(math.sin(a - b), math.cos(a - b))


class HelloSquare(Node):
    """Drive a 1 m x 1 m square using odometry to close the loop."""

    def __init__(self) -> None:
        super().__init__('hello_square')
        self._cmd_pub = self.create_publisher(TwistStamped, 'cmd_vel', 10)
        self._odom_sub = self.create_subscription(
            Odometry, 'platform/odom', self._on_odom, 10
        )
        self._side = 0
        self._phase = 'forward'
        self._have_odom = False
        self._x = 0.0
        self._y = 0.0
        self._yaw = 0.0
        self._start_x = 0.0
        self._start_y = 0.0
        self._start_yaw = 0.0
        self._timer = self.create_timer(0.05, self._tick)
        self.get_logger().info('Hello, Clearpath! Driving a 1m x 1m square.')

    def _on_odom(self, msg: Odometry) -> None:
        self._x = msg.pose.pose.position.x
        self._y = msg.pose.pose.position.y
        self._yaw = yaw_from_quaternion(msg.pose.pose.orientation)
        if not self._have_odom:
            self._start_x = self._x
            self._start_y = self._y
            self._start_yaw = self._yaw
            self._have_odom = True

    def _publish(self, linear: float = 0.0, angular: float = 0.0) -> None:
        msg = TwistStamped()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.twist.linear.x = linear
        msg.twist.angular.z = angular
        self._cmd_pub.publish(msg)

    def _distance_traveled(self) -> float:
        dx = self._x - self._start_x
        dy = self._y - self._start_y
        return math.hypot(dx, dy)

    def _heading_change(self) -> float:
        return abs(angle_diff(self._yaw, self._start_yaw))

    def _tick(self) -> None:
        if not self._have_odom:
            return

        if self._phase == 'forward':
            if self._distance_traveled() + POSITION_TOLERANCE_M < SIDE_LENGTH_M:
                self._publish(linear=LINEAR_SPEED)
            else:
                self._publish()
                self.get_logger().info(f'Side {self._side + 1} complete.')
                self._phase = 'turn'
                self._start_yaw = self._yaw
        elif self._phase == 'turn':
            if self._heading_change() + HEADING_TOLERANCE_RAD < TURN_ANGLE_RAD:
                self._publish(angular=ANGULAR_SPEED)
            else:
                self._publish()
                self._side += 1
                if self._side >= 4:
                    self.get_logger().info('Square complete.')
                    rclpy.shutdown()
                    return
                self._phase = 'forward'
                self._start_x = self._x
                self._start_y = self._y


def main(args=None) -> None:
    """Spin the ``HelloSquare`` node."""
    rclpy.init(args=args)
    node = HelloSquare()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

File: demos/hello_clearpath/src/hello_square.cpp
// Copyright (c) 2026, Clearpath Robotics, by Rockwell Automation, and
// community contributors.
// Licensed under the Apache License, Version 2.0.

#include <cmath>
#include <memory>

#include <geometry_msgs/msg/twist_stamped.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <rclcpp/rclcpp.hpp>

namespace
{
constexpr double kSideLengthM = 1.0;
constexpr double kTurnAngleRad = M_PI / 2.0;
constexpr double kLinearSpeed = 0.2;   // m/s
constexpr double kAngularSpeed = 0.5;  // rad/s
constexpr double kPositionToleranceM = 0.02;
constexpr double kHeadingToleranceRad = 2.0 * M_PI / 180.0;
constexpr auto kTickPeriod = std::chrono::milliseconds(50);

double yaw_from_quaternion(const geometry_msgs::msg::Quaternion & q)
{
  const double siny_cosp = 2.0 * (q.w * q.z + q.x * q.y);
  const double cosy_cosp = 1.0 - 2.0 * (q.y * q.y + q.z * q.z);
  return std::atan2(siny_cosp, cosy_cosp);
}

double angle_diff(double a, double b) { return std::atan2(std::sin(a - b), std::cos(a - b)); }
}  // namespace

class HelloSquare : public rclcpp::Node
{
public:
  HelloSquare() : rclcpp::Node("hello_square")
  {
    cmd_pub_ = create_publisher<geometry_msgs::msg::TwistStamped>("cmd_vel", 10);
    odom_sub_ = create_subscription<nav_msgs::msg::Odometry>(
      "platform/odom", 10, [this](nav_msgs::msg::Odometry::ConstSharedPtr msg) { on_odom(*msg); });
    timer_ = create_wall_timer(kTickPeriod, [this] { tick(); });
    RCLCPP_INFO(get_logger(), "Hello, Clearpath! Driving a 1m x 1m square.");
  }

private:
  enum class Phase
  {
    kForward,
    kTurn
  };

  void on_odom(const nav_msgs::msg::Odometry & msg)
  {
    x_ = msg.pose.pose.position.x;
    y_ = msg.pose.pose.position.y;
    yaw_ = yaw_from_quaternion(msg.pose.pose.orientation);
    if (!have_odom_)
    {
      start_x_ = x_;
      start_y_ = y_;
      start_yaw_ = yaw_;
      have_odom_ = true;
    }
  }

  void publish_cmd(double linear, double angular)
  {
    geometry_msgs::msg::TwistStamped msg;
    msg.header.stamp = now();
    msg.twist.linear.x = linear;
    msg.twist.angular.z = angular;
    cmd_pub_->publish(msg);
  }

  double distance_traveled() const { return std::hypot(x_ - start_x_, y_ - start_y_); }

  double heading_change() const { return std::abs(angle_diff(yaw_, start_yaw_)); }

  void tick()
  {
    if (!have_odom_)
    {
      return;
    }

    if (phase_ == Phase::kForward)
    {
      if (distance_traveled() + kPositionToleranceM < kSideLengthM)
      {
        publish_cmd(kLinearSpeed, 0.0);
      }
      else
      {
        publish_cmd(0.0, 0.0);
        RCLCPP_INFO(get_logger(), "Side %d complete.", side_ + 1);
        phase_ = Phase::kTurn;
        start_yaw_ = yaw_;
      }
    }
    else
    {
      if (heading_change() + kHeadingToleranceRad < kTurnAngleRad)
      {
        publish_cmd(0.0, kAngularSpeed);
      }
      else
      {
        publish_cmd(0.0, 0.0);
        ++side_;
        if (side_ >= 4)
        {
          RCLCPP_INFO(get_logger(), "Square complete.");
          rclcpp::shutdown();
          return;
        }
        phase_ = Phase::kForward;
        start_x_ = x_;
        start_y_ = y_;
      }
    }
  }

  rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr cmd_pub_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
  rclcpp::TimerBase::SharedPtr timer_;
  Phase phase_{Phase::kForward};
  int side_{0};
  bool have_odom_{false};
  double x_{0.0};
  double y_{0.0};
  double yaw_{0.0};
  double start_x_{0.0};
  double start_y_{0.0};
  double start_yaw_{0.0};
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<HelloSquare>());
  rclcpp::shutdown();
  return 0;
}
